### PR TITLE
Add support to install a specific version of plugin

### PR DESCRIPTION
**Pull request by sbcd90**
Added as an option to `install` command

```
jcli plugin install localization-zh-cn jacoco@2.0.1
```
This will install `latest version of localization-zh-cn` & `version 2.0.1 of jacoco plugin`.

### CODE REVIEW

**Comment by LinuxSuRen** [comment]
@sbcd90 Thanks for your contribution. But there is something that you might not know before.
We generate the release notes base on the PR title. So you probable need to adjust it.
Second, every PR should pass the [sonar quality gate](https://sonarcloud.io/dashboard?branch=pr-211&id=jenkins-zh_jenkins-cli).

Sorry about that no PR template clear this. Hopefully, you can understand this.

**Comment by LinuxSuRen** [comment]
I've created the PR template file in #212

**Comment by sbcd90** [comment]
Hi @LinuxSuRen , thanks a lot for these details. But when i see the sonar quality gate results, i dont find the test cases that i have already written in `client/pluginManager_test.go`.

This should cover the uncovered lines shown in sonar quality gate.

image

Can you plz let me know how can i retrigger the sonar build including `client/pluginManager_test.go`?

**Comment by LinuxSuRen** [comment]
@sbcd90 It might be a fault which belongs to Sonarqube. You can make some changes then push it again. Or some people just close this PR, and wait for a couple of minutes, then reopen it. The build can be triggered again.

File: client/pluginManger.go [hunk 1]
```diff
@@ -133,6 +139,53 @@ func (p *PluginManager) InstallPlugin(names []string) (err error) {
...
+
+// InstallPluginWithVersion install a plugin by name & version
+func (p *PluginManager) installPluginWithVersion(name string) (err error) {
+	pluginName := "%s.hpi"
+	pluginVersion := strings.Split(name, "@")
+	url := fmt.Sprintf("http://updates.jenkins-ci.org/download/plugins/%s/%s/%s.hpi",
+		pluginVersion[0], pluginVersion[1], pluginVersion[0])
+
+	if resp, err := http.Get(url); err != nil {
+		fmt.Println(err)
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: If there're some errors happen. I think it cannot keep going.  We probably should just return the error.

File: client/pluginManger.go [hunk 2]
```diff
@@ -133,6 +139,53 @@ func (p *PluginManager) InstallPlugin(names []string) (err error) {
...
+		err = fmt.Errorf("unexpected status code: %d", statusCode)
+		if p.Debug {
+			ioutil.WriteFile(debugLogFile, body, 0664)
+		}
+		return err
+	}
+	
+	if err = ioutil.WriteFile(fmt.Sprintf("%s.hpi", name), body, 0644); err != nil {
+		if p.Debug {
+			ioutil.WriteFile(debugLogFile, []byte(err.Error()), 0664)
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: Consider putting these error handles into a method. So we don't need to offer test cases for everyone.

File: client/pluginManger.go [hunk 3]
```diff
@@ -133,6 +139,48 @@ func (p *PluginManager) InstallPlugin(names []string) (err error) {
...
+
+	err = ioutil.WriteFile(fmt.Sprintf("%s.hpi", name), body, 0644)
+	return err
+}
+
+// InstallPluginWithVersion install a plugin by name & version
+func (p *PluginManager) installPluginWithVersion(name string) (err error) {
+	pluginName := "%s.hpi"
+	pluginVersion := strings.Split(name, "@")
+	url := fmt.Sprintf("http://updates.jenkins-ci.org/download/plugins/%s/%s/%s.hpi",
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: We cannot deal with the dependency of the plugin in this way. So, there're two suggestions below:

- resolve the dependency issues
- giving the warn tip about this

Comment (LinuxSuRen) [reply client/pluginManger.go]: There's a download method that exists. See also https://github.com/jenkins-zh/jenkins-cli/blob/master/client/pluginApi.go#L105

**Comment by LinuxSuRen** [comment]
@sbcd90 Actually, you don't need to force push your commits if you try to avoid too many commits. Because I can squash and merge it. Second, I cannot see your changes after the force pushing. Plus, please let me know if you finished your changes.

File: client/pluginManger.go (hunk 3 above)
Comment (sbcd90) [reply client/pluginManger.go]: @LinuxSuRen , i'm using the `download` api now.

**Comment by sbcd90** [comment]
@LinuxSuRen , i'm done with my changes. i just didnt cover the `upload` part as it is part of a different test case.
Please let me know if you want further changes.

File: client/pluginManger.go [hunk 4]
```diff
@@ -104,20 +106,24 @@ func (p *PluginManager) GetPlugins() (pluginList *InstalledPluginList, err error
...
 
-func getPluginsInstallQuery(names []string) string {
+func (p *PluginManager) getPluginsInstallQuery(names []string) string {
 	pluginNames := make([]string, 0)
 	for _, name := range names {
 		if name == "" {
 			continue
 		}
+		if strings.Contains(name, "@") {
+			p.installPluginWithVersion(name)
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: Function `getPluginsInstallQuery` is only expected to get a query string array. We should not do more things. It'll be more difficult to maintain if it has too many powers. So my suggestion is that do it outside of this function.

File: client/pluginManger.go [hunk 5]
```diff
@@ -104,20 +106,24 @@ func (p *PluginManager) GetPlugins() (pluginList *InstalledPluginList, err error
...
+		}
 		pluginNames = append(pluginNames, fmt.Sprintf("plugin.%s=", name))
 	}
 	return strings.Join(pluginNames, "&")
 }
 
 // InstallPlugin install a plugin by name
 func (p *PluginManager) InstallPlugin(names []string) (err error) {
-	api := fmt.Sprintf("/pluginManager/install?%s", getPluginsInstallQuery(names))
+	api := fmt.Sprintf("/pluginManager/install?%s", p.getPluginsInstallQuery(names))
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: Maybe you can create a function named `getVersionalPlugins` here to install and upload them.

File: client/pluginManger.go [hunk 6]
```diff
@@ -133,6 +139,26 @@ func (p *PluginManager) InstallPlugin(names []string) (err error) {
...
+	pluginName := "%s.hpi"
+	pluginVersion := strings.Split(name, "@")
+	url := fmt.Sprintf("http://updates.jenkins-ci.org/download/plugins/%s/%s/%s.hpi",
+		pluginVersion[0], pluginVersion[1], pluginVersion[0])
+
+	statusCode := pluginAPI.download(url, name)
+
+	if statusCode >= 200 && statusCode < 300 {
+		if err = p.Upload(fmt.Sprintf(pluginName, name)); err != nil {
+			return err
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: The plugin file will not deletes if there's anything goes wrong here. Usually, we download a temporary file into a temporary directory.

File: client/pluginManger.go [hunk 7]
```diff
@@ -133,6 +139,26 @@ func (p *PluginManager) InstallPlugin(names []string) (err error) {
...
 
+// InstallPluginWithVersion install a plugin by name & version
+func (p *PluginManager) installPluginWithVersion(name string) (err error) {
+	pluginAPI := PluginAPI{}
+	pluginName := "%s.hpi"
+	pluginVersion := strings.Split(name, "@")
+	url := fmt.Sprintf("http://updates.jenkins-ci.org/download/plugins/%s/%s/%s.hpi",
+		pluginVersion[0], pluginVersion[1], pluginVersion[0])
+
+	statusCode := pluginAPI.download(url, name)
```
Comment (LinuxSuRen) [inline client/pluginManger.go]: You can delete it by the below way:
`defer os.Remove(fmt.Sprintf(pluginName, name))`

**Comment by sbcd90** [comment]
Hi @LinuxSuRen , thanks for your review. I have made all the changes. Can you plz re-check again?

**Pull request merged**